lib/code-caller-docker.js

```diff
@@ -0,0 +1,685 @@
+        });
+    }
+
+    async _ensureDirectoryIfNeeded() {
+        await this.ensureDirectoryMutex.runExclusive(async () => {
+            if (config.workersExecutionMode === 'native' && !this.hostDirectory) {
+                this.hostDirectory = await tmp.dir({ unsafeCleanup: true, prefix: MOUNT_DIRECTORY_PREFIX });
+            }
+        });
+    }
+
+    /**
+     * Ensures that the required Docker image is present on this machine.
+     * NOTE: in production, this will block code execution if the image isn't already
+     * present locally. Every effort to be made to ensure that the current image has
+     * already been pulled before PrairieLearn starts up.
+     */
+    async _ensureImage() {
+        const image = docker.getImage(getExecutorImageName());
+        try {
+            await image.inspect();
+        } catch (e) {
+            if (e.statusCode === 404) {
+                await docker.pull(getExecutorImageName(), {});
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Creates a container and attaches its stdin/stdout/stderr to streams
+     * we can write to and read from.
+     * 
+     * @param {string[]} binds - The bind mounts for the container
+     */
+    async _createAndAttachContainer(binds) {
+        this.container = await docker.createContainer({
+            name: `prairielearn.worker.${this.uuid}`,
+            Image: getExecutorImageName(),
+            OpenStdin: true,
+            // This will close stdin once we disconnect, which will let
+            // the worker know that they should die. Once the worker dies,
+            // `AutoRemove: true` below will ensure the worker container
+            // is removed.
+            StdinOnce: true,
+            AttachStdin: true,
+            AttachStdout: true,
+            AttachStderr: true,
+            HostConfig: {
+                Binds: binds,
+                AutoRemove: true,
+            },
+        });
+
+        const stream = await this.container.attach({
+            stream: true,
+            stdin: true,
+            stdout: true,
+            stderr: true,
+        });
+        this.stdinStream = new MemoryStream();
+        this.stdoutStream = new MemoryStream();
+        this.stderrStream = new MemoryStream();
+        this.stdinStream.pipe(stream);
+        this.container.modem.demuxStream(stream, this.stdoutStream, this.stderrStream);
+        this.stdoutStream.on('data', data => this._handleOutput(data));
+        // To see stderr from the container for debugging, uncomment the following line
+        // this.stderrStream.pipe(process.stderr);
+
+        await this.container.start();
+
+        // This will passively wait for the container to exit or be killed
+        this.container
+            .wait()
+            .then((status) => this._handleContainerExit(null, status))
+            .catch((err) => this._handleContainerExit(err));
```

Comment:
Should we write this as a `try` wrapped around an `await`? Or is there some advantage to the explicit promise usage?

Reply:
See the comment above, I could have made that a bit clearer. The key is that it *doesn't* `await` the promise here, but allows other code to keep executing. The container should almost never, if ever, die; this is just to handle it gracefully at some arbitrary point in the future if it *does*.

Reply:
I’m not sure which is the “comment above”? But yes, your explanation makes total sense. I think I was just being rather too enthusiastic about wanting to see a/a everywhere 😄

Reply:
Ah, the comment preceding this line of code. "Passively" refers to the fact that it's a non-blocking wait. I should inline my longer explanation in said comment 🙂